lib/ansible/module_utils/vmware.py

```diff
@@ -32,6 +35,117 @@
     HAS_PYVMOMI = False
 
 
+class PyVmomiHelper(object):
+
+    def __init__(self, module):
+        if not HAS_PYVMOMI:
+            module.fail_json(msg='pyvmomi module required')
+
+        self.module = module
+        self.params = module.params
+        self.content = connect_to_api(self.module)
+
+    @staticmethod
+    def construct_vm_searchpath(datacenter, folder):
+        # https://pubs.vmware.com/vi3/sdk/ReferenceGuide/vim.SearchIndex.html#findByInventoryPath
+        #   "My Folder/My Datacenter/vm/Discovered VM/VM1"
+        #   * leading slash or trailing slash is ignored
+        #   * the following paths all represents the same object:
+        #       "a/b", "/a/b", "a/b/", and '/a/b/'
+        #   * Slashes in names must be represented using %2f
+
+        if not folder:
+            folder = '/vm'
+        else:
+            if not folder.startswith('/vm') and not folder.startswith('vm'):
+                folder = os.path.join('/vm', folder)
+
+        if folder.endswith('/'):
+            folder = folder.rstrip('/')
+
+        searchpath = os.path.join(datacenter, folder)
+        return searchpath
+
+    def getvm(self, name=None, uuid=None, searchpath=None):
+        '''Find a VM by name and optionally a folder with absolute certainty'''
+
+        vm = None
+
+        # prefer a UUID
+        if uuid is not None:
+            vm = find_vm_by_id(self.content, uuid, vm_id_type="uuid", instanceUuid=False)
+            return vm
+
+        # use a searchpath + FindByInventoryPath to search by whatever
+        # folder the user thinks it is in.
+        elif searchpath:
+
+            # get all objects for this path ...
+            f_obj = self.content.searchIndex.FindByInventoryPath(searchpath)
+
+            # find the VM object(s)
+            if f_obj:
+                if isinstance(f_obj, vim.Datacenter):
+                    f_obj = f_obj.vmFolder
+                for c_obj in f_obj.childEntity:
+                    if not isinstance(c_obj, vim.VirtualMachine):
+                        continue
+                    if c_obj.name == name:
+                        vm = c_obj
+                        if self.params['name_match'] == 'first':
+                            break
+
+        # If no vm was found with the UUID or the explicit searchpath given,
+        # collect all VMs by the given name and correlate based on their folder
+        # paths.
+        if not vm:
+            vms = find_vm_by_name(
+                self.content,
+                name,
+                folder=None,
+                recurse=True,
+                allow_duplicates=True
+            )
+            if vms:
+                # return if only one name because we don't have to worry
+                # about duplicates
+                if len(vms) == 1:
+                    return vms[0]
+                else:
+                    # need to find the best path for the duplicate VMs
```

Comment:
I appreciate the friendliness of this, but as a user if I've got more than one VM named ``foo`` and I can't give you the right path to it, I think that should fail as a reality check for me.